sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala

```diff
@@ -165,6 +165,24 @@ object ResolveHints {
           hintErrorHandler.hintRelationsNotFound(h.name, h.parameters, unmatchedIdents)
           applied
         }
+      case With(child, relations) => resolveCTEHint(child,
+        relations.foldLeft(Seq.empty[(String, LogicalPlan)]) {
+        case (resolved, (name, relation)) =>
+          resolved :+ name -> apply(resolveCTEHint(relation, resolved))
+      })
+    }
+
+    def resolveCTEHint(plan: LogicalPlan, cteRelations: Seq[(String, LogicalPlan)]): LogicalPlan = {
+      plan resolveOperatorsDown {
+        case u: UnresolvedRelation =>
+          cteRelations.find(x => resolver(x._1, u.tableName)).map(_._2).getOrElse(u)
```

Comment:
This branch will occur `stackoverflow` when cte table name is same as table in cte as follows:
sql("create temporary view t as select 1 as id")      
sql("with **t** as (select /*+ BROADCAST(id) */ id from **t**) select id from t")
@cloud-fan  Could you please help me find a way to pass this ?

Reply:
Have you looked at https://github.com/apache/spark/pull/29062 ? Seems easier to just run CTE substitution in the very beginning.

Reply:
Yes,  `put the CTE substitution in the very beginning` is an easier way.